src/core/reference/include/openvino/reference/roi_align.hpp

```diff
@@ -15,7 +15,189 @@ namespace ov {
+        T y2 = (rois[coordinate_index({index, 3}, shape)] + offset_src) * spatial_scale + offset_dst;
+
+        T roi_width = x2 - x1;
+        T roi_height = y2 - y1;
+
+        if (!aligned) {
+            roi_width = std::max(roi_width, static_cast<T>(1.0));
+            roi_height = std::max(roi_height, static_cast<T>(1.0));
+        }
+
+        return {x1, y1, roi_width, roi_height};
+    }
+
+private:
+    const T* rois;
+    Shape shape;
+    float spatial_scale;
+    bool aligned;
+    T offset_src;
+    T offset_dst;
+};
+
 template <typename T>
+class ROIAlignRotatedOpDefPolicy {
+public:
+    class ROIAlignRotatedSamplingSpace : public SamplingSpaceBase<T> {
+    public:
+        Point<T> transform_sampling_point_to_image_space(const Point<T>& point) const {
+            float y = point.y * cos_angle - point.x * sin_angle + center_y;
+            float x = point.y * sin_angle + point.x * cos_angle + center_x;
+            return {x, y};
+        }
+
+    private:
+        friend class ROIAlignRotatedOpDefPolicy<T>;
+        ROIAlignRotatedSamplingSpace(T center_x_,
+                                     T center_y_,
+                                     float cos_angle_,
+                                     float sin_angle_,
+                                     T start_x_,
+                                     T start_y_,
+                                     T size_x_,
+                                     T size_y_)
+            : SamplingSpaceBase<T>(start_x_, start_y_, size_x_, size_y_),
+              center_x(center_x_),
+              center_y(center_y_),
+              cos_angle(cos_angle_),
+              sin_angle(sin_angle_) {}
+
+        T center_x;
+        T center_y;
+        float cos_angle;
+        float sin_angle;
+    };
+
+    ROIAlignRotatedOpDefPolicy() {}
+
+    void init(const T* rois_, const Shape& shape_, float spatial_scale_, AlignedMode aligned_mode, bool clockwise_) {
+        rois = rois_;
+        shape = shape_;
+        spatial_scale = spatial_scale_;
+        clockwise = clockwise_;
+
+        if (aligned_mode != AlignedMode::ASYMMETRIC) {
+            OPENVINO_THROW(std::string("ROIAlignRotated: Not supported aligned_mode"));
+        }
+    }
+
+    ROIAlignRotatedSamplingSpace get_sampling_space_for_index(unsigned int index) {
+        const T center_x = (rois[coordinate_index({index, 0}, shape)]) * spatial_scale - 0.5f;
+        const T center_y = (rois[coordinate_index({index, 1}, shape)]) * spatial_scale - 0.5f;
+        const T width = (rois[coordinate_index({index, 2}, shape)]) * spatial_scale;
+        const T height = (rois[coordinate_index({index, 3}, shape)]) * spatial_scale;
+        T angle = (rois[coordinate_index({index, 4}, shape)]);
+
+        if (clockwise) {
+            angle = -angle;
+        }
+        const float cos_angle = cos(angle);
+        const float sin_angle = sin(angle);
+
+        const T x1 = -width / 2.0f;
+        const T y1 = -height / 2.0f;
```

Comment:
I was trying to follow, but I'm not sure why the negative of width and height is used here.

Reply:
This is a very good question.
Short answer is: to have a base of local coordinate system in the center of the box (because you want to do the rotation around box's center). You could set your base anywhere, but this one is the most convenient one in this case IMO. If this does not satisfy you, let's make a video call, will explain it deeply.